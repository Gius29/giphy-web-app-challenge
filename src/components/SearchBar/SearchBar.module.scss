.searchBar {
  button, input {
      padding: 5px 10px;
      border-radius: 5px;
      border: 1px solid #ddd;
      margin: 5px;
      font-size: 20px;
    }
    button {
      width: 120px;
    }
    input{
      width: 360px;
    }
}
