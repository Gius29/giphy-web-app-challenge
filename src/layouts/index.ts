// Each constant represent a possible layout that the app may have.
export enum LayoutIDs {
  ONE_COLUMN = "ONE_COLUMN",
  THREE_COLUMNS = "THREE_COLUMNS"
}
