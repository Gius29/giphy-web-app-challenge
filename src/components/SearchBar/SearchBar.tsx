import React, { Component } from "react";
import { LayoutIDs } from "../../layouts";

import styles from "./SearchBar.module.scss";

interface StateProps {
  placeHolderMsg: string;
  layoutID: LayoutIDs;
}

interface DispatchProps {
  searchQuery: (query: string) => void;
  changeLayout: () => void;
}

type SearchBarProps = StateProps & DispatchProps;

interface OwnState {
  query: string;
}

class SearchBar extends Component<SearchBarProps, OwnState> {
  constructor(props: SearchBarProps) {
    super(props);

    this.state = { query: "" };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ query: event.target.value });
  }

  handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    this.props.searchQuery(this.state.query);
    event.preventDefault();
  }

  selectLayoutMessage() {
    const { layoutID } = this.props;
    if (layoutID === LayoutIDs.ONE_COLUMN) {
      return "3 columns";
    }
    // LayoutIDs.THREE_COLUMNS (There are no other layout possible)
    return "1 column";
  }

  render() {
    const { placeHolderMsg } = this.props;
    const layoutMessage = this.selectLayoutMessage();

    return (
      <form className={styles.searchBar} onSubmit={this.handleSubmit}>
        <input
          value={this.state.query}
          placeholder={placeHolderMsg}
          onChange={this.handleChange}
        />
        <button type="submit">Search</button>
        <button type="button" onClick={this.props.changeLayout}>
          {layoutMessage}
        </button>
      </form>
    );
  }
}

export default SearchBar;
