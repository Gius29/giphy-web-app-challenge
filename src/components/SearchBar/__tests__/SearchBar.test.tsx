import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { LayoutIDs } from "../../../layouts";
import SearchBar from "../SearchBar";

describe("SearchBar", () => {
  const mockPlaceHolderMsg = "mockPlaceHolderMsg";

  it("should renders correctly with ONE_COLUMN layout", () => {
    const wrapper = renderer.create(
      <SearchBar
        placeHolderMsg={mockPlaceHolderMsg}
        layoutID={LayoutIDs.ONE_COLUMN}
        searchQuery={jest.fn()}
        changeLayout={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should renders correctly with THREE_COLUMNS layout", () => {
    const wrapper = renderer.create(
      <SearchBar
        placeHolderMsg={mockPlaceHolderMsg}
        layoutID={LayoutIDs.THREE_COLUMNS}
        searchQuery={jest.fn()}
        changeLayout={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should update the query field and perform a search", () => {
    const mockSearchQuery = jest.fn();

    const wrapper = mount(
      <SearchBar
        placeHolderMsg={mockPlaceHolderMsg}
        layoutID={LayoutIDs.ONE_COLUMN}
        searchQuery={mockSearchQuery}
        changeLayout={jest.fn()}
      />
    );

    //Interactions:
    // 1 - Simulate an imput form the user
    // 2 - simulate submit by the user
    wrapper
      .find("input")
      .simulate("change", { target: { value: "newQueryString" } });
    wrapper.find('[type="submit"]').simulate("submit");

    wrapper.unmount();
    expect(mockSearchQuery).toHaveBeenCalledTimes(1);
    expect(mockSearchQuery).toHaveBeenCalledWith("newQueryString");
  });
});
