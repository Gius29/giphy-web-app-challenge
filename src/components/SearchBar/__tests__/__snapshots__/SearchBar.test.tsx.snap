// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should renders correctly with ONE_COLUMN layout 1`] = `
<form
  className="searchBar"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="mockPlaceHolderMsg"
    value=""
  />
  <button
    type="submit"
  >
    Search
  </button>
  <button
    onClick={[MockFunction]}
    type="button"
  >
    3 columns
  </button>
</form>
`;

exports[`SearchBar should renders correctly with THREE_COLUMNS layout 1`] = `
<form
  className="searchBar"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="mockPlaceHolderMsg"
    value=""
  />
  <button
    type="submit"
  >
    Search
  </button>
  <button
    onClick={[MockFunction]}
    type="button"
  >
    1 column
  </button>
</form>
`;
